add resource quotas

ivelichkovich opened the pull request:
refactored version of this https://github.com/Praqma/helmsman/pull/340, we're going to use helm3 and helmsman3 so don't need tiller resources so just resource quotas

luisdavim commented on internal/app/kube_helpers.go:
```diff
@@ -23,6 +23,7 @@ func addNamespaces(namespaces map[string]namespace) {
 			labelNamespace(name, cfg.Labels)
 			annotateNamespace(name, cfg.Annotations)
 			setLimits(name, cfg.Limits)
+			setQuotas(nsName, ns.Quotas)
```
this is actually what is causing the test to fail:
```
internal/app/kube_helpers.go:26:14: loop variable nsName captured by func literal
internal/app/kube_helpers.go:26:22: loop variable ns captured by func literal
```

ivelichkovich replied:
Yeah just saw that, my bad

ivelichkovich commented on internal/app/kube_helpers.go:
```diff
@@ -134,6 +135,64 @@ spec:
+spec:
+  hard:
+`
+	if quotas.Pods != "" {
```
I just copy pasted what I had from the old PR before helm 3, will change

ivelichkovich replied:
Looking at it, the marshalling won't work because it isn't the exact same object format because of the customResources option

luisdavim replied:
what matters is the format we use in the Helmsman file and the struct format we use to unmarshal, we can make them match, like it's done in the `setLimits` function....

ivelichkovich replied:
No we can't because of the customResources field, that can be anything, like gpu's or something custom, so we can't have that because the struct would need to be dynamic

ivelichkovich replied:
requests.nvidia.com/gpu I guess is a standard one, although not very common so I didn't include it but you can have resource quotas on completely custom resources such as count/widgets.example.com or wheels/custom.com

luisdavim replied:
Sure, but all the other parts can be marshaled and this appended at the end like you are doing now.

ivelichkovich replied:
so two questions

1, how would I tell it to marshal everything but the customResources?

2, this is what the actual yaml looks like for k8s
status:
  hard:
    limits.cpu: "10"
    limits.memory: 8Gi
    pods: "25"
    requests.cpu: "10"
    requests.memory: 8Gi 

So I would have to change the structure of the helmsman yaml would have to change, we can do that but I think it'd be a little weird to have the dot notation

ivelichkovich replied:
@luisdavim hey this should be resolved now, sorry my day job was very busy this last month. Let me know if the way I handled not marshaling the custom quotas is okay

ivelichkovich commented on internal/app/kube_helpers.go:
```diff
@@ -134,6 +135,64 @@ spec:
+		log.Fatal(err.Error())
+	}
+
+	cmd := command{
```
didn't realize we had one, copy paste again, fixed

ivelichkovich commented on docs/how_to/namespaces/quotas.md:
````diff
@@ -0,0 +1,45 @@
+        value: 0
+```
+
+The example above will create two namespaces - staging and production - with resource limits defined for the staging namespace.
````
fixed

mkubaczyk commented on internal/app/kube_helpers.go:
```diff
@@ -134,6 +135,50 @@ spec:
+		log.Fatal("ERROR: failed to create ResourceQuota in namespace [ " + ns + " ]: " + result.errors)
+	}
+
+	deleteFile("temp-ResourceQuota.yaml")
```
I'd rather put this line before we run log.Fatal for result.code != 0 so the file "always" gets deleted even when kubectl apply failed

mkubaczyk replied:
Also you might want to use this catalog https://github.com/Praqma/helmsman/blob/master/internal/app/main.go#L10 for such file

mkubaczyk commented on docs/how_to/namespaces/quotas.md:
```diff
@@ -0,0 +1,44 @@
+---
+version: ???
```
You can set it as 3.2.0, I hope this will go with next minor release

luisdavim commented:
Hi @ivelichkovich would you be able to address @mkubaczyk's comments?
Thanks.

ivelichkovich commented:
Yes working on it now

ivelichkovich commented:
Updated per the requests and I added quotas into the examples/example.yaml,
 I'm not sure how to create tests for the setQuotas method without a big refactor of the rest of the code because I can't mock out the call to kubectl or the log.fatal, if there's a better way to mock or any ideas I'd be happy to implement them

mkubaczyk commented on docs/how_to/namespaces/quotas.md:
```diff
@@ -0,0 +1,44 @@
+---
+version: 3.2.0
```
It took us so long so 3.2.0 was actually released. Let's change it to 3.3.0 and we can merge it :-)

ivelichkovich replied:
Sure, sorry about that, work has been very busy

ivelichkovich replied:
Updated :)
